### DISCUSSION TITLE
What should I do to redirect a serial device (/dev/ttyUSB0) under ubuntu20 host to a virtual machine windows 10 system?

### DISCUSSION THREAD

**Discussion by fan4905** [virt-manager/virt-manager#780] [Q&A]
Please, I want to give the serial device (/dev/ttyUSB0) of the host (ubuntu20 system) to the windows10 inside the virtual machine, I configure the place in the virtualization above virt-manager, add hardware, select the serial port, the type is psuedo TTY (pty), and then start the virtual machine into the I can see the com port in windows 10 device manager, but it doesn't work, how can I associate it with my host serial device /dev/ttyUSB0?
Thank you!please reply me!

**Comment by phrdina** [comment]
Hi, if the USB serial device is visible in your host as USB device (you can check `lsusb`) in your VM you should be able to attach the whole USB device to the VM by clicking `Add Hardware`, selecting `USB Host Device` and there you would pick your USB serial device.

If that doesn't work you can also try the following command:
`virt-xml VMNAME --add-device --serial type=dev,source.path=/dev/ttyS4,target.type=usb-serial`.

Converting to discussion as it is not an issue.

**Reply by fan4905** [reply]
Hello, I may not have described it clearly. I am a serial device, not a USB device, and I have added a USB to serial port connector. If used as a USB device, the COM device can be seen in the Windows Device Manager of the virtual machine. However, when I open this COM device with software, it prompts that it cannot be used. That is to say, USB to serial devices cannot be added as USB devices. Have I described it clearly?

**Reply by fan4905** [reply]
Addendum: virt-xml VMNAME -- add device -- serial type=dev, source. path=/dev/ttyS4, target. type=usb serial. This command can add a serial device, but the com port cannot be seen in the device manager of Windows in the virtual machine.

**Reply by phrdina** [reply]
Did you change the path from /dev/ttyS4 to /dev/ttyUSB0 ?

**Reply by fan4905** [reply]
> Did you change the path from /dev/ttyS4 to /dev/ttyUSB0 ?

Yes, it has been changed. The command I used was: virt-xml VMNAME -- add device -- serial type=dev, source. path=/dev/ttyS0， As a result, when entering the Windows Virtual Machine Device Manager, the COM port device cannot be seen

**Reply by fan4905** [reply]
![image](https://github.com/user-attachments/assets/01597c86-d0dc-4ede-bd8c-fdf8030a8e72)
![image](https://github.com/user-attachments/assets/1fbc709b-994d-4ab2-917f-d1e542fd44b5)
COM devices cannot be used
